src/minimetrics/core.py

```diff
@@ -0,0 +1,424 @@
+                if not force_flush and ts > cutoff:
+                    continue
+
+                extracted_metric: ExtractedMetric = {
+                    "type": ty,
+                    "name": name,
+                    "value": metric.serialize_value(),
+                    "timestamp": ts,
+                    "width": int(self.ROLLUP_IN_SECONDS),
+                }
+                if unit:
+                    extracted_metric["unit"] = unit
+                if tags:
+                    extracted_metric["tags"] = tags
+
+                extracted_metrics.append((extracted_metric, metric.current_complexity))
+                flushed_buckets.add(bucket_key)
+                complexity_to_remove += metric.current_complexity
+
+            # We remove all flushed buckets, in order to avoid memory leaks.
+            for bucket_key in flushed_buckets:
+                buckets.pop(bucket_key)
+
+            self._force_flush = False
+            self._bucket_complexity -= complexity_to_remove
+
+        if extracted_metrics:
+            self._emit(extracted_metrics, force_flush)
+
+    def add(
+        self,
+        ty: MetricType,
+        key: str,
+        value: MetricValue,
+        unit: MetricUnit,
+        tags: Optional[MetricTagsExternal],
+        timestamp: Optional[float],
+    ) -> None:
+        if self._flusher is None:
+            return
+
+        if timestamp is None:
+            timestamp = time.time()
+
+        bucket_key: ComposedKey = (
+            int((timestamp // self.ROLLUP_IN_SECONDS) * self.ROLLUP_IN_SECONDS),
+            ty,
+            key,
+            unit,
+            # We have to convert tags into our own internal format, since we don't support lists as
+            # tag values.
+            self._to_internal_metric_tags(tags),
+        )
+
+        with self._lock:
+            metric = self.buckets.get(bucket_key)
+            if metric is None:
+                metric = METRIC_TYPES[ty]()
+                self.buckets[bucket_key] = metric
+
+            # We first change the complexity by taking the old one and the new one.
+            previous_complexity = metric.current_complexity
+            metric.add(value)
+            self._bucket_complexity += metric.current_complexity - previous_complexity
+            # Given the new complexity we consider whether we want to force flush.
+            self.consider_force_flush()
+
+    def stop(self):
+        if self._flusher is None:
+            return
+
+        # Firstly we tell the flusher that we want to force flush.
+        with self._lock:
+            self._force_flush = True
+            self._running = False
+
+        # Secondly we notify the flusher to move on and we wait for its completion.
+        self._flush_event.set()
+        self._flusher.join()
+        self._flusher = None
+
+    def consider_force_flush(self):
+        total_complexity = len(self.buckets) + self._bucket_complexity
```

Comment:
This formula has a `+` not a `*` and was just drafted as a naive estimate. In order to properly implement it we should give a weight to each component since in the best case `len(self.buckets)` will be equal to `self._bucket_complexity` and we might want to weigh differently the amount of buckets w.r.t. the complexity.

Reply:
I see, thanks!